vscode/src/chat/chat-view/SimpleChatPanelProvider.ts

```diff
@@ -597,170 +529,236 @@ export class SimpleChatPanelProvider implements vscode.Disposable, ChatSession {
+        try {
+            const prompt = await this.buildPrompt(prompter, sendTelemetry)
+            this.streamAssistantResponse(requestID, prompt)
+        } catch (error) {
+            if (isRateLimitError(error)) {
+                this.postError(error, 'transcript')
+            } else {
+                this.postError(isError(error) ? error : new Error(`Error generating assistant response: ${error}`))
+            }
+        }
     }
 
     /**
-     * Handles a chat request from chat input or a command.
-     *
-     * Saves the chat session, posts a transcript update, generates the
-     * assistant's response, logs telemetry, and updates the panel title.
+     * Constructs the prompt and updates the UI with the context used in the prompt.
      */
-    private async handleChatRequest(
-        requestID: string,
-        inputText: string,
-        submitType: 'user' | 'suggestion' | 'example' | 'user-newchat',
-        userContextFiles: ContextFile[],
-        addEnhancedContext: boolean,
-        command?: CodyCommand
-    ): Promise<void> {
-        // Display text is the text we will display to the user in the Chat UI
-        // - Append @-files to the display text if we have any
-        // - Append @-file selection for commands
-        // Otherwise, use the input text
-        const displayText = userContextFiles?.length
-            ? createDisplayTextWithFileLinks(inputText, userContextFiles)
-            : command
-              ? createDisplayTextWithFileSelection(
-                      inputText,
-                      this.editor.getActiveTextEditorSelectionOrEntireFile()
-                  )
-              : inputText
-        // The text we will use to send to LLM
-        const promptText = command ? command.prompt : inputText
-        this.chatModel.addHumanMessage({ text: promptText }, displayText)
+    private async buildPrompt(prompter: IPrompter, sendTelemetry?: (contextSummary: {}) => void): Promise<Message[]> {
+        const { prompt, contextLimitWarnings, newContextUsed } = await prompter.makePrompt(
+            this.chatModel,
+            getContextWindowForModel(this.authProvider.getAuthStatus(), this.chatModel.modelID)
+        )
 
-        await this.saveSession(inputText)
-        // trigger the context progress indicator
-        this.postViewTranscript({ speaker: 'assistant' })
-        await this.generateAssistantResponse(
-            requestID,
-            userContextFiles,
-            addEnhancedContext,
-            contextSummary => {
-                if (submitType !== 'user' && submitType !== 'user-newchat') {
-                    return
+        // Update UI based on prompt construction
+        this.chatModel.setNewContextUsed(newContextUsed)
+        if (contextLimitWarnings.length > 0) {
+            const warningMsg = contextLimitWarnings
+                .map(w => (w.trim().endsWith('.') ? w.trim() : w.trim() + '.'))
```

Comment:
Minor nit but I think the following is more readable. The purpose is obscured by calling 'trim' three times. Even better if we have a helper which does this already and can just call it.

```suggestion
                .map(w => {
                  w = w.trim();
                  if (!w.endsWith('.')) {
                    w += '.'
                  }
                  return w
                })
```